### ISSUE TITLE
perf: Many goroutine stacks need to grow

### ISSUE THREAD

**Issue by bdarnell** [cockroachdb/cockroach#8544]
On the beta cluster running 78f6418, we see almost 15% of CPU time spent in the `runtime.morestack` function (see attached profile), indicating that goroutines are growing beyond the default 4KB limit (and possibly beyond the subsequently increased limit). This is a relatively expensive process due to the pointer adjustments that must be made. 

Unfortunately, go's profiling support does not tell us where these growing stacks are coming from, so it's hard to track this down. If we have any large structs allocated on the stack it may be worthwhile to move them to a sync.Pool (or even just a plain allocation - allocating and garbage collecting a 4KB buffer should be cheaper than a 4KB stack allocation. 

[profile.pdf](https://github.com/cockroachdb/cockroach/files/418535/profile.pdf)

**Comment by petermattis** [comment]
I'd be hesitant to make changes to what we put on the stack without understanding where the stack growth is coming from. We could hack something in which allowed us to determine the size of a goroutine stack via a call. Specifically, [petermattis/goid](https://github.com/petermattis/goid)  shows how to hack access to the internal `g` structure. I believe [`g.stackAlloc`](https://github.com/petermattis/goid/blob/master/goid_go1.5plus.go#L56) is the size of the current goroutine's stack.

**Comment by bdarnell** [comment]
I believe the stack size of a function is static at compile time. Google had some tooling for C++ to find functions that required large stack allocations. If the culprit is a single allocation (rather than something like recursion), the same thing should be possible for Go.

**Comment by petermattis** [comment]
@bdarnell Yeah, I could have sworn I saw a tool like that for Go as well, but I couldn't find it in my Google searches just now. The `g.stackAlloc` approach is something we could do fairly quickly.

**Comment by petermattis** [comment]
We're doing a lot of stack allocation per `Replica.handleRaftReady` invocation and each invocation is on a fresh goroutine. The goroutine stack starts at 2KB. I can see that grow (via doubling) to 32KB. Looks like that problematic code path is when we're executing an `EndTransaction`, though I haven't traced through where all of the stack usage is occurring.

**Comment by petermattis** [comment]
`go build -gcflags=-f` outputs compiler debug information about stack frames in the package being built. For example, here is a snippet from `Replica.handleRaftReady`:

```
--- (*Replica).handleRaftReady frame ---
NAME r G2 *Replica width=8
NAME ~r0 G1 error width=16
NAME rd G5 raft.Ready width=264
NAME autotmp_2743 G0 raftpb.Message width=256
NAME msg G12 raftpb.Message width=256
NAME command G15 roachpb.RaftCommand width=144
NAME command G22 roachpb.RaftCommand width=144
```

The receiver `r` takes 8 bytes of stack space and the return value `error` takes 16 (an interface is 2 pointers).

We store the raft ready struct on the stack which consumes 264 bytes. Looping over the outgoing messages we temporarily store each one on the stack which consumes 256 bytes. 

It isn't clear at what stage this debug output is kept or how good a job the Go compiler does with liveness analysis. 

One item this debug output reveals is our habit of passing structs by value even when a pointer is readily available has a cost. For example, `Replica.executeCmd` passes the request by value to the various implementation methods which requires copying the value to the stack even though we have a pointer readily available. 

I don't think there is short term work to be done here, though. I think we'd be better off restructuring `Store.processRaft` to reuse the goroutines executing `Replica.handleRaftReady`.

**Comment by tamird** [comment]
This is resolved by #8574, afaik.
